src/account_manager/builder.rs

```diff
@@ -154,53 +156,63 @@ impl AccountManagerBuilder {
-                        self.coin_type
-                            .ok_or(crate::Error::MissingParameter("coin_type (IOTA: 4218, Shimmer: 4219)"))?,
-                    )
-                }
+                None => (
+                    self.client_options
+                        .clone()
+                        .ok_or(crate::Error::MissingParameter("client_options"))?,
+                    self.secret_manager
+                        .clone()
+                        .ok_or(crate::Error::MissingParameter("secret_manager"))?,
+                    self.coin_type
+                        .ok_or(crate::Error::MissingParameter("coin_type (IOTA: 4218, Shimmer: 4219)"))?,
+                ),
             };
 
-            let client = client_options.clone().finish()?;
+            self.client_options.replace(client_options.clone());
+            self.secret_manager.replace(secret_manager.clone());
+            self.coin_type.replace(coin_type);
 
-            let accounts = storage_manager.lock().await.get_accounts().await.unwrap_or_default();
+            // Store account manager data in storage
+            storage_manager.lock().await.save_account_manager_data(&self).await?;
+
+            let client = client_options.clone().finish()?;
 
             #[cfg(feature = "events")]
             let event_emitter = Arc::new(Mutex::new(EventEmitter::new()));
 
+            let mut accounts = storage_manager.lock().await.get_accounts().await.unwrap_or_default();
+
+            let mut account_handles: Vec<AccountHandle> = accounts
+                .into_iter()
+                .map(|a| {
+                    AccountHandle::new(
+                        a,
+                        client.clone(),
+                        secret_manager.clone(),
+                        #[cfg(feature = "events")]
+                        event_emitter.clone(),
+                        storage_manager.clone(),
+                    )
+                })
+                .collect::<_>();
+
+            // If the manager builder is not set, it means the user provided it and we need to update the addresses.
+            // In the other case it was loaded from the database and addresses are up to date.
+            if manager_builder.is_none() {
+                for account in account_handles.iter_mut() {
+                    account.update_account_with_new_client(client.clone()).await?;
+                }
+            }
+
             return Ok(AccountManager {
                 #[cfg(not(feature = "events"))]
-                accounts: Arc::new(RwLock::new(
-                    accounts
-                        .into_iter()
-                        .map(|a| AccountHandle::new(a, client.clone(), secret_manager.clone(), storage_manager.clone()))
-                        .collect(),
-                )),
+                accounts: Arc::new(RwLock::new(account_handles)),
                 #[cfg(feature = "events")]
```

Comment:
`#[cfg(feature = "events")]` and `#[cfg(not(feature = "events"))]` can be removed here, no?

Reply:
line 215 we pass the event_emitter, so we need the `#[cfg(feature = "events")]` in this case, don't we?

Reply:
good catch, fixed it